### ISSUE TITLE
Add support for viewing and editing protobuf content using Google's protoc package

### ISSUE THREAD

**Issue by aarnaut** [mitmproxy/mitmproxy#5001]
#### Problem Description
As of now, mitmproxy only supports raw deserialization of protobuf data. There is no way to provide a set of proto files in order to get "contextual" deserialization. Editing the content is not possible either, as it  requires proto definition in order to serialize the content.

#### Proposal
I've built a few addons using mitmproxy's public API. The implementation is working based on descriptor files and is running on top of [google's protobuf package](https://googleapis.dev/python/protobuf/latest/) . See [this repo](https://github.com/aarnaut/mitmproxy-grpc) for more info. I've only tested it with the services I use, so this is still experimental. My proposal is to add these addons to the list of default ones and expose it to a wider audience. 

#### Alternatives
An alternative would be to update the existing protobuf implementation that is available in mitmproxy. I found that difficult, as the current one is an implementation of a protobuf serializer and deserializer. I don't have the context of why it was done like this, but on first glance, I find it better to use packages that provide implementation of the protobuf spec, as it requires minimal maintenance. But perhaps there was a good reason to go with a custom implementation.

**Comment by mame82** [comment]
The customizable gRPC view (with protobuf support) was created for two reasons:
- lack of gRPC support
- preference of "flattened" deserialization, to join (nested) field tags across multiple levels and assign custom decoding/naming where needed

Initial idea was to add minimal modification, which means that the Kaitai deserializer should be used. It turned out that Kaitai had minor issues, that's why I moved to manual deserialization.

It was discussed with @mhils that integrating "native" protobuf support would have several advantages. The most important ones are:

- support for proto definitions (incl. future versions)
- serialization support (editing of flow bodies)

There are also shortcomings:

- requires external native dependency (at least when last evaluated)
- a way has to be found to allow users to assign correct proto definitions to message bodies (f.e. choose correct definition based on query path, "best fit" etc etc) and to store multiple definition files
- no support for "out of specs" usage (f.e. nested, packed messages, as seen in Google API traffic)
- more effort for "hot reload" support (file watcher per proto def file, instead of hot reload for a single python add-on, when definitions change)

IMHO it is definitely worth to consider an implementation relying on Google's protoc codebase.

I have no time left to work on this, but left behind some code snippets for serialization in the test cases for the gRPC add-on (only for wire format).

Additional note: Major changes in the wire format by Google, are less likely than changes on the upper encoding layers (where definitions join the game)

**Comment by mame82** [comment]
Related PRs (including related discussions/comments) for reference:

- https://github.com/mitmproxy/mitmproxy/pull/4851
- https://github.com/mitmproxy/mitmproxy/pull/4858
- https://github.com/mitmproxy/mitmproxy/pull/4867
- Kaitai replacement + out-of-specs decoding:
https://github.com/mitmproxy/mitmproxy/pull/4872


Hope this helps, sorry for not being able to offer more support.

**Comment by mame82** [comment]
Also for reference:

GIF animation showing hot reload of content view, while editing "pseudo" proto definitions (protocol reversing style, as definitions are only partially known and thus wouldn't be well-formed).

https://twitter.com/mame82/status/1447981116623888393

**Comment by aarnaut** [comment]
Thanks for such a detailed response! I went through the PRs, it was a lot to unpack. Please correct me if I'm wrong in my understanding.

>There are also shortcomings:
>
> * requires external native dependency (at least when last evaluated)

If you're referring to protoc, it is possible to avoid it if we ask for a compiled descriptor file as an input.

> * a way has to be found to allow users to assign correct proto definitions to message bodies (f.e. choose correct definition based on query path, "best fit" etc etc) and to store multiple definition files

Usually you can resolve the method and message type based on the path. However, there are probably edge cases and it should be possible to allow for custom mapping of `path->message type`.

> * no support for "out of specs" usage (f.e. nested, packed messages, as seen in Google API traffic)

I don't have enough information on this one. I'm wondering if they have any plans of supporting it in case it's still not supported.

> * more effort for "hot reload" support (file watcher per proto def file, instead of hot reload for a single python add-on, when definitions change)

Do we need hot reload support in such cases where you provide the full definition right on start?

The way I see it, the existing gRPC content view is targeting a power user or someone who explicitly wants to work with raw data. It's a custom implementation that supports flattened deserialization and allows for fine-grained tuning of parsing rules. Fine-grained approach works great when you need high level of control, but becomes difficult for an average user (e.g. tester) who just needs to inspect and modify traffic with minimal effort possible. Correct me if I'm wrong, but adding new parsing rules is a lot of manual work, especially if you don't need any custom parsing other than assigning names to field values. This especially becomes an issue when working with a large number of services, as it was in my case.

Relying on the protoc package would simplify things both for us and the user. There would be far less maintenance from our side, especially when it comes to supporting future versions. We don't need to deal with the low level specs of the format. Additionally, we get out of the box functionality that allows us to serialize/deserialize protobuf data as well as resolve gRPC methods and protobuf messages by using descriptor files. For users, it will be more convenient to use, as they only need to provide a descriptor of their `.proto` files. 

Perhaps it would make sense to have both views present; one for more fine-grained control and the other one for all the other average use cases.

In any case, how do you want to proceed here? Do you think it would be worth to creating a PR and having a detailed look at the proposal I have in mind, or do you think it's better to adjust the existing custom implementation?

**Comment by Prinzhorn** [comment]
> The way I see it, the existing gRPC content view is targeting a power user or someone who explicitly wants to work with raw data.
> Fine-grained approach works great when you need high level of control, but becomes difficult for an average user (e.g. tester) who just needs to inspect and modify traffic with minimal effort possible.
> For users, it will be more convenient to use, as they only need to provide a descriptor of their `.proto` files.

Correct me if I'm wrong, but you seem to use mitmproxy to debug/test _your own_ gRPC for which you have access to the `.proto` files. But there is a much larger use-case which is pentesting/reverse engineering endpoints you do _not_ own. And if you don't have `.proto` files, you need to make sense of the raw data you see on the wire. I just want to make sure that this use-case is not overlooked.

Can the official protobuf package even work _without_ the `.proto` file? I think what we currently have is as far as you can get with the raw data, or not? So this would be as separate implementation? But I'm just a user of gRPC/protobuf.

I see two entirely separate feature requests here:

1. Allow editing protobuf in the UI and serialize it back (listed as possible future work in https://github.com/mitmproxy/mitmproxy/pull/4851)
2. Optionally support `.proto` files to make the viewing (and editing) better

**Comment by aarnaut** [comment]
Yes, that's completely right. This is why I also mentioned in in the end:

>Perhaps it would make sense to have both views present; one for more fine-grained control and the other one for all the other average use cases.

>Can the official protobuf package even work without the .proto file? 

Good point. I would have to check this. I don't see why it shouldn't fall back to raw deserialization. In any case, this specific use-case would require `.proto` files anyway.

> I think what we currently have is as far as you can get with the raw data, or not? So this would be a separate implementation? 

Yes, I think the same. The existing raw implementation is targeting a specific audience and it's doing a good job at it. However, it is not very accessible to those that want to test their own gRPC services. I think it would be better to have separate implementations rather than mixing the two use-cases by extending the existing raw deserialization. 

Adding all this support to existing implementation would be a lot of work, and it probably stretches beyond something that a pentester would use, right? I think it would be good to draw a line at some point and basically say - Okay, we have optimized support for flattened raw deserialization with custom rules that a pentester or someone who does reverse engineering would use. Anything beyond that requires context i.e. `.proto` files, and for that we can rely on a third-party package that supports it out of the box.

**Comment by mhils** [comment]
I agree that the reverse engineering and debugging use cases are distinct enough to warrant separate implementations. I don't think we can extend @aarnaut's implementation to support unknown protobufs[^1], at the same time we don't have the resources to add and maintain `.proto` support for @mame82's current implementation. Long story short, I'd be happy to have both in mitmproxy. 

We need to make sure that this is not super confusing to users, but there should be a way. For starters, the .proto contentview could show a helpful "no .proto file configured, configure it or use the other view if you don't have one" message if it can't figure out stuff.

For editing, the custom command is an interesting workaround I'd be willing to accept for now. I think we should do a better job here long-term, but that applies not just to protobuf but to all other binary formats as well.

[^1]: `protoc` has `--decode_raw`, but that is very basic. protoc is also a distinct thing from the PyPI protobuf package, which AFAIK cannot do raw decode.

**Comment by aarnaut** [comment]
Great, I can prepare a PR and we can all take a look. 

> For editing, the custom command is an interesting workaround I'd be willing to accept for now. I think we should do a better job here long-term, but that applies not just to protobuf but to all other binary formats as well.

I've actually implemented another approach that pops up a window and lets you choose an option, similarly to how content view window works. The thing is that it requires user to edit their mitmproxy config. But that shouldn't be required anymore, now that we can integrate it directly with mitmproxy. What do you think about that?

**Comment by mhils** [comment]
> I've actually implemented another approach that pops up a window and lets you choose which content body to edit, similarly to how content view window works. The thing is that it requires user to edit their mitmproxy config. But that shouldn't be required anymore, now that we can integrate it directly with mitmproxy. What do you think about that?

Contentviews are shared between mitmproxy and mitmweb, so we don't want a console integration here really. Eventually I'd like reserialization to be part of the contentviews API itself, but defining that properly will take quite some effort.

**Comment by yoshimo** [comment]
Does this work also take care of grpc Web which is slightly different than the one over http/2? https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md

**Comment by aarnaut** [comment]
I haven't tried gRPC web yet, so no. Feel free to check out the PR and see if you can get it working. Perhaps we don't need many changes.

**Comment by 89z** [comment]
> Can the official protobuf package even work without the `.proto` file?

Not sure about Python, but the Go package can [1]. However since the ProtoBuf wire format is not self-describing, the package doesn't handle ambiguous data very well [2], as a workaround, I wrote my own package [3], that parses Length-delimited types as both string and message.

> protoc has `--decode_raw`, but that is very basic. protoc is also a distinct thing from the PyPI protobuf package, which AFAIK cannot do raw decode.

Have you looked at BlackBoxProtoBuf [4]?


1. https://godocs.io/google.golang.org/protobuf/testing/protopack
2. https://github.com/golang/protobuf/issues/1370
3. https://github.com/89z/format
4. https://github.com/nccgroup/blackboxprotobuf

**Comment by mhils** [comment]
This will be fixed by https://github.com/mitmproxy/mitmproxy/pull/7623.
